session.go

```diff
@@ -340,126 +330,76 @@ func (s *Session) keepalive() {
 	}
 }
 
-// waitForSendErr waits to send a header, checking for a potential shutdown
-func (s *Session) waitForSend(hdr header, body io.Reader) error {
-	errCh := make(chan error, 1)
-	return s.waitForSendErr(hdr, body, errCh, nil)
-}
-
-func pooledTimer(d time.Duration) (*time.Timer, func()) {
-	t := timerPool.Get()
-	timer := t.(*time.Timer)
-	timer.Reset(d)
-	cancelFn := func() {
-		timer.Stop()
-		select {
-		case <-timer.C:
-		default:
-		}
-		timerPool.Put(t)
-	}
-	return timer, cancelFn
-}
-
-// waitForSendErr waits to send a header with optional data, checking for a
-// potential shutdown. Since there's the expectation that sends can happen
-// in a timely manner, we enforce the connection write timeout here.
-//
-// timeout is:
-//  * `nil` if (a) this is a control message (ping, go away, window update, etc.), or (b) if the user
-//    has not set a write deadline on the stream.
-//  * non-`nil` if (a) this is a user-requested write, and (b) the stream has a write deadline.
-func (s *Session) waitForSendErr(hdr header, body io.Reader, errCh chan error, timeout <-chan time.Time) error {
+// send sends the header and body.
+func (s *Session) sendMsg(hdr header, body []byte, deadline <-chan struct{}) error {
 	select {
 	case <-s.shutdownCh:
-		return ErrSessionShutdown
+		return s.shutdownErr
 	default:
 	}
 
-	var connWriteTimerCh <-chan time.Time
-
-	if timeout == nil {
-		// fall back to the connection write timeout.
-		t, cancelFn := pooledTimer(s.config.ConnectionWriteTimeout)
-		defer cancelFn()
-		connWriteTimerCh = t.C
-	}
-
-	ready := &sendReady{Hdr: hdr, Body: body, Err: errCh, Stage: stageInitial}
+	// duplicate as we're sending this async.
+	buf := pool.Get(headerSize + len(body))
+	copy(buf[:headerSize], hdr[:])
+	copy(buf[headerSize:], body)
```

Comment:
This is the most annoying part of this patch for me. However:

1. To implement per-stream deadlines correctly, we _need_ to copy the data before writing. Otherwise, we can't return while the write is still in-progress (we don't want to just kill the entire connection).
2. The buffer pool is actually pretty fast.

Reply:
yeah, it's ok.